agent/dns.go

```diff
@@ -1483,10 +1369,284 @@ func findWeight(node structs.CheckServiceNode) int {
+	return []dns.RR{ipRecord}, nil
+}
+
+// Craft dns records for an IP
+// In case of an SRV query the answer will be a IN SRV and additional data will store an IN A to the IP
+// Otherwise it will return a IN A record
+func (d *DNSServer) makeRecordFromIP(dc string, addr string, serviceNode structs.CheckServiceNode, req *dns.Msg, ttl time.Duration) ([]dns.RR, []dns.RR) {
+	q := req.Question[0]
+	ipRecord := makeARecord(q.Qtype, addr, ttl)
+	if ipRecord == nil {
+		return nil, nil
+	}
+
+	if q.Qtype == dns.TypeSRV {
+		ipFQDN := d.encodeIPAsFqdn(dc, addr)
+		answers := []dns.RR{
+			&dns.SRV{
+				Hdr: dns.RR_Header{
+					Name:   q.Name,
+					Rrtype: dns.TypeSRV,
+					Class:  dns.ClassINET,
+					Ttl:    uint32(ttl / time.Second),
+				},
+				Priority: 1,
+				Weight:   uint16(findWeight(serviceNode)),
+				Port:     uint16(serviceNode.Service.Port),
+				Target:   ipFQDN,
+			},
+		}
+
+		ipRecord.Header().Name = ipFQDN
+		return answers, []dns.RR{ipRecord}
+	}
+
+	ipRecord.Header().Name = q.Name
+	return []dns.RR{ipRecord}, nil
+}
+
+// Craft dns records for an FQDN
+// In case of an SRV query the answer will be a IN SRV and additional data will store an IN A to the IP
+// Otherwise it will return a CNAME and a IN A record
+func (d *DNSServer) makeRecordFromFQDN(dc string, fqdn string, serviceNode structs.CheckServiceNode, req *dns.Msg, ttl time.Duration, cfg *dnsConfig, maxRecursionLevel int) ([]dns.RR, []dns.RR) {
+	edns := req.IsEdns0() != nil
+	q := req.Question[0]
+
+	more := d.resolveCNAME(cfg, dns.Fqdn(fqdn), maxRecursionLevel)
+	var additional []dns.RR
+	extra := 0
+MORE_REC:
+	for _, rr := range more {
+		switch rr.Header().Rrtype {
+		case dns.TypeCNAME, dns.TypeA, dns.TypeAAAA:
+			// set the TTL manually
+			rr.Header().Ttl = uint32(ttl / time.Second)
+			additional = append(additional, rr)
+
+			extra++
+			if extra == maxRecurseRecords && !edns {
+				break MORE_REC
+			}
+		}
+	}
+
+	if q.Qtype == dns.TypeSRV {
+		answers := []dns.RR{
+			&dns.SRV{
+				Hdr: dns.RR_Header{
+					Name:   q.Name,
+					Rrtype: dns.TypeSRV,
+					Class:  dns.ClassINET,
+					Ttl:    uint32(ttl / time.Second),
+				},
+				Priority: 1,
+				Weight:   uint16(findWeight(serviceNode)),
+				Port:     uint16(serviceNode.Service.Port),
+				Target:   dns.Fqdn(fqdn),
+			},
+		}
+		return answers, additional
+	}
+
+	answers := []dns.RR{
+		&dns.CNAME{
+			Hdr: dns.RR_Header{
+				Name:   q.Name,
+				Rrtype: dns.TypeCNAME,
+				Class:  dns.ClassINET,
+				Ttl:    uint32(ttl / time.Second),
+			},
+			Target: dns.Fqdn(fqdn),
+		}}
+	answers = append(answers, additional...)
+
+	return answers, nil
+}
+
+func (d *DNSServer) nodeServiceRecords(dc string, node structs.CheckServiceNode, req *dns.Msg, ttl time.Duration, cfg *dnsConfig, maxRecursionLevel int) ([]dns.RR, []dns.RR) {
+	nodeAddrIsIP := net.ParseIP(node.Node.Address) != nil
+
+	// There is no service address and the node address is an IP
+	if node.Service.Address == "" && nodeAddrIsIP {
```

Comment:
Note for release/1-6 branch integration: We now have service level tagged addresses and those should be handled here.

There are a few other places in this function where we look at node.Service.Address. We should probably call TranslateAddress once and store the results and just use that everywhere instead.

Reply:
Done